File: day2/day2.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

const (
	Forward = iota
	Up
	Down
)

type Movement struct {
	Direction int
	Magnitude int
}

type Coordinates struct {
	x int
	y int
}

func readMovements(file string) []Movement {
	body, err := ioutil.ReadFile(file)
	if err != nil {
		fmt.Println(err)
	}

	lines := strings.Split(string(body), "\n")

	var movement *Movement
	var movements []Movement
	var number int

	for _, line := range lines {
		movement = new(Movement)

		fields := strings.Fields(line)
		number, err = strconv.Atoi(string(fields[1]))
		if err != nil {
			fmt.Println(err)
		}
		movement.Magnitude = number

		switch fields[0] {
		case "forward":
			movement.Direction = Forward
		case "down":
			movement.Direction = Down
		case "up":
			movement.Direction = Up
		}

		movements = append(movements, *movement)
	}

	return movements
}

func calculatePosition(movements []Movement) Coordinates {
	var coordinates Coordinates
	for _, movement := range movements {
		switch movement.Direction {
		case 0:
			coordinates.x += movement.Magnitude
		case 1:
			coordinates.y -= movement.Magnitude
		case 2:
			coordinates.y += movement.Magnitude
		}
	}

	return coordinates
}

func main() {
	movements := readMovements("day2.txt")
	coordinates := calculatePosition(movements)
	fmt.Println(coordinates.x * coordinates.y)
}

File: day2/day2.hs
import Data.List (foldl')
import Text.Read (readMaybe)

data Direction = Forward | Up | Down deriving (Show, Eq)

data Movement = Movement { direction :: Direction, magnitude :: Int } deriving (Show, Eq)

main :: IO ()
main = readMovements >>= print . calculatePosition

directionFrom :: String -> Maybe Direction
directionFrom "forward" = Just Forward
directionFrom "up" = Just Up
directionFrom "down" = Just Down
directionFrom _ = Nothing

parseMovement :: [String] -> Maybe Movement
parseMovement (x:y:_) = do
    d <- directionFrom x
    m <- readMaybe y :: Maybe Int
    pure $ Movement d m

readMovements :: IO [Maybe Movement]
readMovements = readFile "day2.txt" >>=
    return . map parseMovement . map words . lines

calculatePosition :: [Maybe Movement] -> Int
calculatePosition = calculate . position
    where
        position = foldl' (\(x, y, z) move -> case move of
                Just (Movement Forward m) -> (x + m, y + z * m, z)
                Just (Movement Up m) -> (x, y, z - m)
                Just (Movement Down m) -> (x, y, z + m)
            ) (0, 0, 0)
        calculate (x, y, _) = x * y
